### PR TITLE
virtualenv: Bump version to 20.0.13 and add dependencies

### PR DESCRIPTION
**Pull request by rubendibattista**
Add py-distlib port that is a depenency of virtualenv

Fix this bug: https://github.com/pypa/virtualenv/issues/1581

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.2
Xcode 11.3.1

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: python/py-virtualenv/Portfile [hunk 1]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
         PortGroup   deprecated 1.0
         deprecated.eol_version  yes
     }
 
-    depends_lib-append  port:py${python.version}-setuptools
+    depends_lib-append  port:py${python.version}-setuptools \
+                        port:py${python.version}-importlib-metadata \
+                        port:py${python.version}-filelock \
+                        port:py${python.version}-distlib \
+                        port:py${python.version}-appdirs
```
Comment (mf2k) [inline python/py-virtualenv/Portfile]: Please sort the dependencies alphabetically.

**Comment by mf2k** [comment]
The commit message needs to be modified. Ideally this would be 2 commits. One for the new port, which needs to be mentioned in this commit. And the other commit for "py-virtualenv" not "virtualenv" as the current CM calls it.

File: python/py-distlib/Portfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+name                py-distlib
+version             0.3.0
+revision            0
+maintainers         {@rubendibattista gmail.com:rubendibattista} openmaintainer
+platforms           darwin
+supported_archs     noarch
+description         Low-level components of distutils2/packaging, augmented \
+                    with higher-level APIs for making packaging easier.
+long_description    ${description}
```
Comment (reneeotten) [inline python/py-distlib/Portfile]: please remove trailing whitespace

File: python/py-distlib/Portfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+platforms           darwin
+supported_archs     noarch
+description         Low-level components of distutils2/packaging, augmented \
+                    with higher-level APIs for making packaging easier.
+long_description    ${description} 
+homepage            https://bitbucket.org/pypa/distlib
+
+license             PSF
+use_zip             yes
+
```
Comment (reneeotten) [inline python/py-distlib/Portfile]: please remove trailing whitespace

File: python/py-virtualenv/Portfile [hunk 4]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
 master_sites        pypi:v/${python.rootname}
 distname            ${python.rootname}-${version}
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: you can remove the ```master_sites``` and ```distname``` lines, they are set by default in the python PG

File: python/py-virtualenv/Portfile [hunk 5]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
         PortGroup   deprecated 1.0
         deprecated.eol_version  yes
     }
 
-    depends_lib-append  port:py${python.version}-setuptools
+     depends_lib-append     port:py${python.version}-appdirs \
+                            port:py${python.version}-distlib \
+                            port:py${python.version}-filelock \
+                            port:py${python.version}-importlib-metadata \
+                            port:py${python.version}-setuptools
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: please remove trailing whitespace on this line

File: python/py-virtualenv/Portfile [hunk 6]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
+                    sha256  7e4e234e1f27755685dc54063d989756790b0682d60e304db589fa1604938013 \
+                    size    7972183
 
 # keep older Python versions here, do add the EOL version to the list below
 python.versions     27 34 35 36 37 38
 
 if {${name} ne ${subport}} {
     # add EOL warnings
-    if {${python.version} in "34"} {
+    if {${python.version} in {"27" "34"}} {
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: this can become ```if {${python.version} in "27 34"} {```, better readability in my opinion (what you might work as well, I haven't checked).

File: python/py-virtualenv/Portfile [hunk 7]
```diff
@@ -5,7 +5,7 @@ PortGroup           python 1.0
 PortGroup           select 1.0
 
 name                py-virtualenv
-version             16.7.9
+version             20.0.2
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: since your PR the port has been updated to 20.0.3, perhaps do this update within this PR as well?

File: python/py-virtualenv/Portfile [hunk 8]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
     # add EOL warnings
-    if {${python.version} in "34"} {
+    if {${python.version} in {"27" "34"}} {
         PortGroup   deprecated 1.0
         deprecated.eol_version  yes
     }
 
-    depends_lib-append  port:py${python.version}-setuptools
+     depends_lib-append     port:py${python.version}-appdirs \
+                            port:py${python.version}-distlib \
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: you appear to be missing a few dependencies: 
- ```six``` for all Python versions
- ```contextlib2``` for Python < 3.3
- ```importlib-resources``` for Python < 3.7
- ```pathlib2``` for Python < 3.4

please check under ```install_requires``` in upstream's ```setup.cfg``` for the released version.

Comment (reneeotten) [reply python/py-virtualenv/Portfile]: additionally, you'll need to add subports for PY34 back to the new dependencies of ```virtualenv```

File: python/py-virtualenv/Portfile [hunk 9]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
-    post-destroot {
-        set docdir ${prefix}/share/doc/${subport}
-        xinstall -d ${destroot}${docdir}/docs/changelog
-        xinstall -m 0644 -W ${worksrcpath} README.rst LICENSE.txt AUTHORS.txt \
-            ${destroot}${docdir}
-        xinstall -m 0644 {*}[glob -directory ${worksrcpath}/docs *.rst] \
-                ${destroot}${docdir}/docs
-        xinstall -m 0644 {*}[glob -directory ${worksrcpath}/docs/changelog *] \
-                ${destroot}${docdir}/docs/changelog
-    }
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: please keep the ```post-destroot``` block, just update it so that whatever files there are in the new version get installed

File: python/py-virtualenv/Portfile [hunk 10]
```diff
@@ -22,38 +22,31 @@ homepage            https://virtualenv.pypa.io
...
 if {${name} ne ${subport}} {
     # add EOL warnings
-    if {${python.version} in "34"} {
+    if {${python.version} in {"27" "34"}} {
         PortGroup   deprecated 1.0
         deprecated.eol_version  yes
     }
 
-    depends_lib-append  port:py${python.version}-setuptools
+     depends_lib-append     port:py${python.version}-appdirs \
```
Comment (reneeotten) [inline python/py-virtualenv/Portfile]: your whitespacing/alignment isn't correct here

File: python/py-virtualenv/Portfile (hunk 8 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Does `python34` need to be supported? I think it would be better to remove support for `python27` and support `python >= 3.5`.

File: python/py-virtualenv/Portfile (hunk 9 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: I checked and there wasn't any files that needed to get installed.

File: python/py-virtualenv/Portfile (hunk 8 above)
Comment (reneeotten) [reply python/py-virtualenv/Portfile]: Yes, that decision was  made a long time ago (there have been comments about this on the mailinglist and/or other PRs, cannot find them right now). Port like ```virtualenv```, ```pip```, ```setuptools``` should keep supporting Python versions that are present in MacPorts (i.e., in lang/pythonXY) so that they can be used to install packages (from PyPI) even though these themselves are not present in MacPorts anymore.

File: python/py-virtualenv/Portfile (hunk 9 above)
Comment (reneeotten) [reply python/py-virtualenv/Portfile]: There is  a ```LICENSE``` and ```README.md``` file and also the ```docs``` directory is still present. So there isn't a real need to change this, or do you find them not useful? I would say there is no harm in installing the first two files, but perhaps the ```.rst``` documentation files and stuff in the ```changelog``` directory aren't all that useful for an end-user. What do you think?

Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: I tried to do `sudo port destroot py37-virtualenv`, and these files were not present. I'm doing something wrong? 

```
$> find /opt/local/var/macports/build/_Users_rubendibattista_git_macports-ports_python_py-virtualenv/py37-virtualenv/work/destroot -type f | grep README
```

No output. Same for `LICENSE`. Same for:
```
find /opt/local/var/macports/build/_Users_rubendibattista_git_macports-ports_python_py-virtualenv/py37-virtualenv/work/destroot -type d | grep docs
```

Comment (reneeotten) [reply python/py-virtualenv/Portfile]: If you don't add the ```post-destroot``` block the file(s) will indeed not be present in the ```destroot``` directory (they're copied in that block to the ```destroot``` dir). 

I would do ```sudo port extract py37-virtualenv``` to fetch/extract the archive that will be used for the installation, the you can do ```cd port `dir py37-virtualenv`/work``` and ```cd``` into the directory that is present there. You can check which files are present there and might be useful to install in the ```post-destroot``` block. Additionally, you'll find the ```setup.py``` and/or ```setup.cfg``` file that you can use double-check the dependencies. Hope that makes sense?

File: python/py-virtualenv/Portfile (hunk 8 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Very clear. I thought that `port destroot` would stop after destroot (and before post-destroot). 

Thanks!

File: python/py-virtualenv/Portfile (hunk 9 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Very clear. I thought that port destroot would stop after destroot (and before post-destroot).

Thanks!

**Comment by rubendibattista** [comment]
The new version `20.0.3` gives an error about `__version__.py` file. I think they need to fix it.

```
:info:build AttributeError: module 'virtualenv' has no attribute '__version__'
:info:build Command failed:  cd "/opt/local/var/macports/build/_Users_rubendibattista_git_macports-ports_python_py-virtualenv/py37-virtualenv/work/virtualenv-20.0.3" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build -j8
```

File: python/py-virtualenv/Portfile [hunk 11]
```diff
@@ -19,24 +19,44 @@ long_description    virtualenv is a tool to create isolated Python \
...
         PortGroup   deprecated 1.0
+
         deprecated.eol_version  yes
     }
 
-    depends_lib-append  port:py${python.version}-setuptools
+    depends_lib-append  port:py${python.version}-appdirs \
+                        port:py${python.version}-distlib \
+                        port:py${python.version}-filelock \
+                        port:py${python.version}-setuptools \
```
Comment (gaborbernat) [inline python/py-virtualenv/Portfile]: doesn't depend on setuptools 👍

File: python/py-virtualenv/Portfile [hunk 12]
```diff
@@ -46,7 +66,7 @@ if {${name} ne ${subport}} {
     post-destroot {
         set docdir ${prefix}/share/doc/${subport}
         xinstall -d ${destroot}${docdir}/docs/changelog
-        xinstall -m 0644 -W ${worksrcpath} README.rst LICENSE.txt AUTHORS.txt \
+        xinstall -m 0644 -W ${worksrcpath} README.md LICENSE \
```
Comment (gaborbernat) [inline python/py-virtualenv/Portfile]: is this file needed? why?

File: python/py-virtualenv/Portfile (hunk 7 above)
Comment (gaborbernat) [inline python/py-virtualenv/Portfile]: update to 20.0.3?

File: python/py-virtualenv/Portfile [hunk 13]
```diff
@@ -19,24 +19,44 @@ long_description    virtualenv is a tool to create isolated Python \
                     environments.
 
 homepage            https://virtualenv.pypa.io
-master_sites        pypi:v/${python.rootname}
-distname            ${python.rootname}-${version}
```
Comment (gaborbernat) [inline python/py-virtualenv/Portfile]: where do you specify the build dependencies from you're purposefully ignoring the other build dependencies? https://github.com/pypa/virtualenv/blob/master/pyproject.toml#L3-L5?

Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Ehi thanks to take time to review this.

I'm working on this... But it looks like `virtualenv` needs `py-setuptools_scm[toml]`. How do I specify in a Portfile the `toml` extra?

File: python/py-virtualenv/Portfile (hunk 12 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: In a previous review @reneeotten made me keep these files. :)

File: python/py-virtualenv/Portfile (hunk 13 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: I guess I just provide the dependency... I'm dumb! Sorry for asking.

File: python/py-virtualenv/Portfile (hunk 11 above)
Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Shouldn't  all the python ports depend on setuptools?

Comment (gaborbernat) [reply python/py-virtualenv/Portfile]: nope 👍 only old-style setup.py builds (which are not supported anymore) so hope we don't do that... if you build via pip or pep-517 this should not be needed

Comment (rubendibattista) [reply python/py-virtualenv/Portfile]: Mh, could you please link me an example of a portfile using `pep-517` or `pip` to install? I tried to grep around but I didn't find anything useful... Or it's automatic with the PG?

Comment (reneeotten) [reply python/py-virtualenv/Portfile]: we do build with ```setup.py``` so the build- and runtime dependency on ```setuptools``` is needed here

File: python/py-setuptools_scm/Portfile [hunk 14]
```diff
@@ -25,7 +25,9 @@ checksums           rmd160  bdddb3b45653392ac098004c2fd9ee166015cd63 \
                     size    45549
 
 if {${name} ne ${subport}} {
-    depends_lib-append  port:py${python.version}-setuptools
+    depends_build-append port:py${python.version}-setuptools \
+                         
+    depends_run-append   port:py${python.version}-toml
```
Comment (reneeotten) [inline python/py-setuptools_scm/Portfile]: It's listed as ```extra``` dependency so not strictly necessary for its functioning. If other ports depend on ```setuptools-scm[toml]``` you could just add a build dependency there on ```py-toml```. On the other hand, it's likely that ports will start using this more and then it could make sense to add it here as default. I am not sure what would be the best.

File: python/py-virtualenv/Portfile (hunk 11 above)
Comment (gaborbernat) [reply python/py-virtualenv/Portfile]: setuptools does not support this behaviour anymore, other than backwards compatible way, so you might probably want to change to pip and wheels 🤷‍♂️

File: python/py-setuptools_scm/Portfile (hunk 14 above)
Comment (gaborbernat) [reply python/py-setuptools_scm/Portfile]: We do use of for building setuptools as we store build logic in there pep 517👍

File: python/py-virtualenv/Portfile (hunk 11 above)
Comment (gaborbernat) [reply python/py-virtualenv/Portfile]: virtualenv itself uses pep 517 and 518 standards to build itself, so I don't think you'll be able to build without provisioning first the build system per pep-518

**Comment by gaborbernat** [comment]
we've now released 20.0.4 👍

**Comment by rubendibattista** [comment]
We should be ok. It still errors on `py34` since `setuptools` is not anymore compatible with `python < 3.5` and it errors out when building. 

What should I do?

**Comment by reneeotten** [comment]
> We should be ok. It still errors on `py34` since `setuptools` is not anymore compatible with `python < 3.5` and it errors out when building.
> 
> What should I do?

You specify that in the case of PY34 it installs the latest version that is compatible (i.e., you'll need a ```if {${python.version} == 34} {}``` block).

**Comment by rubendibattista** [comment]
The failurews don't look related to my edits. I think you can merge!

**Comment by gaborbernat** [comment]
We've released 20.0.5 yesterday, with proper support for macOs framework python 2 support; also https://github.com/pypa/virtualenv/pull/1653 is went in today and will be released with 20.0.6, probably on Monday. I'd recommend holding on with promoting this together with that, and release straight 20.0.6.

**Comment by rubendibattista** [comment]
I'll never be able to delive. 🤣. 

Ok I try to update to 20.0.5 now and I'll wait for 20.0.6 :)

**Comment by gaborbernat** [comment]
I'm sorry the new major release causes so many issues; macOS framework builds seems particularly different than another CPython packaging, so got a few more tries than perhaps should have (then again being an open-source project only so much I can do in my free time).

**Comment by rubendibattista** [comment]
@gaborbernat Ehi, I was joking. Thanks for the work you do! :)

**Comment by rubendibattista** [comment]
I think this is quite important, since the current version has bugs with subprocess? Can we merge this and then, as soon as 20.0.6 will come out, I'll update again?

**Comment by gaborbernat** [comment]
I'd wait for Monday, https://github.com/pypa/virtualenv/pull/1711 will land this weekend with ``20.0.11`` 👍

**Comment by rubendibattista** [comment]
Ok, I updated at the last version available. :). Please merge this since the available version on Macports is bugged.

**Comment by gaborbernat** [comment]
This is now ready to go 👍

**Comment by emcrisostomo** [comment]
@reneeotten your requested changes are outdated. Can you have a look at them to confirm this merge is now OK?

**Comment by reneeotten** [comment]
okay, I will take care of getting this merged. There are still a few minor points left (e.g., ```py-setuptools``` should be upgrade to its latest version and that's a maintained port, so one would need give the maintainer some time to respond). So I will fetch this PR locally, make some minor changes where required, and commit the pieces already that are good-to-go. Once that's done, I will force-push the remaining ports to this PR and we should be able to get this update merged. Sounds okay @rubendibattista ?

**Comment by rubendibattista** [comment]
@reneeotten Ehi, I think it's ok even if I believe that this could land now and with another PR then you could upload setuptools. Otherwise with my and your lags, we will never merge it :)

**Comment by reneeotten** [comment]
> @reneeotten Ehi, I think it's ok even if I believe that this could land now and with another PR then you could upload setuptools. Otherwise with my and your lags, we will never merge it :)

there were still some issues regarding dependencies with this PR, but I'm almost done with it. There is no need to update ```py-setuptools```, so I am just leaving it out; additionally I've squashed the addition of py34-suborts into one commit. Currently testing to see if everything works as intended and then I'll push it to ```master``` and we can close this PR. Stay tuned ;) and thanks for your work!

**Comment by rubendibattista** [comment]
@reneeotten Ok! Thanks for the support!

**Comment by reneeotten** [comment]
I just committed your new/updated ports with a few modifications to make sure everything works correctly. Thanks again for the update and sorry that it took so long this time get this committed!

**Comment by rubendibattista** [comment]
So we can close this PR I guess. Thanks to everybody for helping me!
